File: 145_Binary_Tree_Postorder_Traversal.cpp
// 145_Binary_Tree_Postorder_Traversal 
//算法思想：后根遍历二叉树
//递归的方式

class Solution {
public:

    void postorder(TreeNode *root, vector<int> &result){
        
        if(root != NULL){
            postorder(root->left, result);
            postorder(root->right, result);
            result.push_back(root->val);
        }else{
            return;
        }
    }
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> result;
        postorder(root, result);
        return result;
    }
};

//非递归方式，通过标志位，当起右子树访问过的时候，栈顶元素抛出
class Solution {
public:

    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> result;
        TreeNode *p = root;
        TreeNode *last = NULL;
        stack<TreeNode *> stack;
        
        while(p != NULL || !stack.empty()){
            if(p != NULL){
                stack.push(p);
                p = p->left;
            }else{
                p = stack.top();
                if(p->right == NULL || p->right == last){
                    result.push_back(p->val);
                    last = p; 
                    stack.pop();
                    p = NULL;
                }else{
                    p = p->right;
                }
            }
        }
        
        return result;
    }
};

File: 145_BinaryTreePostorderTraversal .java
/*Given a binary tree, return the postorder traversal of its nodes' values.

For example:
Given binary tree {1,#,2,3},
   1
    \
     2
    /
   3
return [3,2,1].

Note: Recursive solution is trivial, could you do it iteratively?
*/

//算法：后续遍历二叉树，Java实现，递归的方式

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {

    List<Integer> postor(TreeNode root, List<Integer> result) {
        if (root != null) {
            result = postor(root.left, result);
            result = postor(root.right, result);
            result.add(root.val);
        }
        return result;
    }

    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<Integer>();
        result = postor(root, result);
        return result;
    }
}

//非递归的方式，栈顶元素，使用peek()
public class Solution {

    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<Integer>();
        TreeNode p = root;
        TreeNode flag = null;
        Stack<TreeNode> stack = new Stack<TreeNode>();
        while (p != null || !stack.empty()) {
            if (p != null) {
                stack.push(p);
                p = p.left;
            } else {
                p = stack.peek();
                if (p.right == null || p.right == flag) {
                    result.add(p.val);
                    stack.pop();
                    flag = p;
                    p = null;
                } else {
                    p = p.right;
                    flag = p;
                }
            }
        }
        return result;
    }
}
